app/uk/gov/hmrc/iossreturns/services/ReturnsService.scala

```diff
@@ -0,0 +1,146 @@
+      getAllPeriodsBetween(commencementLocalDate, LocalDate.now(clock)).map(period =>
+        getReturn(iossNumber, period).map((period, _))
+      )
+    val vatReturns: Future[Seq[(Period, Option[EtmpVatReturn])]] = Future.sequence(vatReturnsResult)
+    vatReturns.map(periodAndReturns => {
+      val allPeriodAndReturns = periodAndReturns.map(periodAndReturn => {
+        val (period, vatReturn) = periodAndReturn
+        val decision: PeriodWithStatus = decideStatus(period, vatReturn, exclusions)
+        decision
+      })
+      addNextIfAllCompleted(allPeriodAndReturns.toList, commencementLocalDate)
+    })
+  }
+
+  private def addNextIfAllCompleted(currentPeriods: List[PeriodWithStatus], commencementLocalDate: LocalDate): List[PeriodWithStatus] = {
+    val nextPeriod = getNextPeriod(currentPeriods.map(_.period), commencementLocalDate)
+    if (currentPeriods.forall(_.status == SubmissionStatus.Complete)) {
+      currentPeriods ++ Seq(PeriodWithStatus(nextPeriod, SubmissionStatus.Next))
+    } else {
+      currentPeriods
+    }
+  }
+
+  def getNextPeriod(periods: List[Period], commencementLocalDate: LocalDate): Period = {
+    val runningPeriod = Period.getRunningPeriod(LocalDate.now(clock))
+    if (periods.nonEmpty) {
+        periods.maxBy(_.lastDay.toEpochDay).getNext()
+    } else {
+      if (commencementLocalDate.isAfter(runningPeriod.lastDay)) {
+        Period.getRunningPeriod(commencementLocalDate)
+      } else {
+        runningPeriod
+      }
+    }
+
+  }
+  def getAllPeriodsBetween(commencementDate: LocalDate, endDate: LocalDate): List[Period] = {
+    val startPeriod = Period(commencementDate.getYear, commencementDate.getMonth)
+    val endPeriod = Period(endDate.getYear, endDate.getMonth)
+
+    val lastPeriod = if (endDate.isBefore(endPeriod.lastDay))
+      endPeriod.getPrevious()
+    else
+      endPeriod
+
+    getAllPeriodsUntil(Nil, startPeriod, lastPeriod)
+  }
+
+  @tailrec
+  private def getAllPeriodsUntil(periodsUpToNow: List[Period], startPeriod: Period, endPeriod: Period): List[Period] = {
+    if (endPeriod.isBefore(startPeriod))
+      Nil
+    else if (startPeriod == endPeriod)
+      startPeriod :: periodsUpToNow
+    else
+      getAllPeriodsUntil(startPeriod :: periodsUpToNow, startPeriod.getNext(), endPeriod)
+  }
+
+  def decideStatus(period: Period, vatReturn: Option[EtmpVatReturn], exclusions: List[EtmpExclusion]) = {
+    if (isPeriodExcluded(period, exclusions))
+      PeriodWithStatus(period, SubmissionStatus.Excluded)
+    else
+      vatReturn match {
+        case Some(_) => PeriodWithStatus(period, SubmissionStatus.Complete)
+        case None => if (LocalDate.now(clock).isAfter(period.paymentDeadline)) {
+          PeriodWithStatus(period, SubmissionStatus.Overdue)
+        } else {
+          PeriodWithStatus(period, SubmissionStatus.Due)
+        }
+      }
+  }
+
+  def getExclusion(exclusions: List[EtmpExclusion]): Option[EtmpExclusion] = {
```

Comment:
Rename `getLastExclusion` ?
```
def getLastExclusion(exclusions: List[EtmpExclusion]): Option[EtmpExclusion] = 
  exclusions.maxByOption(_.effectiveDate).filterNot(_.exclusionReason == Reversal)
```

Reply:
Thanks, changed to getLastExclusionWithoutReversal to be more precise.